activerecord/lib/active_record/railtie.rb

```diff
@@ -339,5 +339,29 @@ class Railtie < Rails::Railtie # :nodoc:
         end
       end
     end
+
+    initializer "active_record.query_log_tags_config" do |app|
+      config.after_initialize do
+        if app.config.active_record.query_log_tags_enabled
+          ActiveRecord::QueryLogs.taggings.merge! \
+            application:  -> { @application_name ||= Rails.application.class.name.split("::").first },
+            pid:          -> { Process.pid }
+
+          if app.config.active_record.query_log_tags.present?
+            ActiveRecord::QueryLogs.tags = app.config.active_record.query_log_tags
+          end
+
+          if app.config.active_record.cache_query_log_tags
+            ActiveRecord::QueryLogs.cache_query_log_tags = true
+          end
+
+          ActiveSupport.on_load(:active_record) do
+            ConnectionAdapters::AbstractAdapter.descendants.each do |klass|
+              klass.prepend(QueryLogs::ExecutionMethods) if klass.descendants.empty?
+            end
```

Comment:
I'm afraid this might be a bit too brittle.

For context we use https://github.com/Shopify/activerecord-pedantmysql2-adapter, which is a subclass of `MySQL2Adapter`. This means `MySQL2Adapter` isn't tagged at all in our app.

We also used to have a fake adapter for test purpose that was causing trouble.

I also feel that having to patch both `exec_query` and `execute` is a bit of a smell, there must be a better way we can hook in there with a single entry point. @kaspth any ideas?

Reply:
Hmm interesting! I would have expected the class to be picked up:

```ruby
irb(main):009:0> ActiveRecord::Base.connection.class
=> ActiveRecord::ConnectionAdapters::PedantMysql2Adapter
irb(main):010:0> ActiveRecord::ConnectionAdapters::AbstractAdapter.descendants
=> [ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter, ActiveRecord::ConnectionAdapters::Mysql2Adapter, ActiveRecord::ConnectionAdapters::PedantMysql2Adapter]
irb(main):011:0> ActiveRecord::ConnectionAdapters::PedantMysql2Adapter.descendants.empty?
=> true
irb(main):012:0> ActiveRecord::QueryLogs.tags = [ :application ]
irb(main):013:0> ActiveRecord::QueryLogs.add_query_log_tags_to_sql("")
=> " /*application:MysqlDevApp*/"
```

I can't remember exactly why these two entry points were required - I think both are the bare minimum used across multiple adapters for different types of query, but not 100% sure. Would love to bring that down to a single place though!

On holiday right now (hah:), can dig more if needed later tonight :)

Reply:
> On holiday right now (hah:), can dig more if needed later tonight :)

Don't worry. I'll have a look. Enjoy your vacations!

Reply:
So my main problem happens because the in the MySQLAdapter `exec_query` calls `execute_and_free` which calls `execute`.

I assume that's also why there's these `return if sql.include?(comment)`, that really sound like a big smell to me.

I'll try to work on refactoring that to eliminate that double call.